analyses/cms-open-data-ttbar/ml.py

```diff
@@ -0,0 +1,297 @@
+            .Define(f'{feature_names[16]}', 'Take(Jet_qgl_cut, W1_idx)')
+            .Define(f'{feature_names[17]}','Take(Jet_qgl_cut, W2_idx)')
+            .Define(f'{feature_names[18]}', 'Take(Jet_qgl_cut, bH_idx)')
+            .Define(f'{feature_names[19]}', 'Take(Jet_qgl_cut, bL_idx)')
+
+    #         # four - momentumes
+
+            .Define(
+                'JetW1_p4', 
+                f'ConstructP4({feature_names[8]}, JetW1_eta, JetW1_phi, Take(Jet_mass_cut, W1_idx))'
+            )
+            .Define(
+                'JetW2_p4', 
+                f'ConstructP4({feature_names[9]}, JetW2_eta, JetW2_phi, Take(Jet_mass_cut, W2_idx))' 
+            )
+            .Define(
+                'JetbL_p4', 
+                f'ConstructP4({feature_names[11]}, JetbL_eta, JetbL_phi, Take(Jet_mass_cut, bL_idx))' 
+            )
+            .Define(
+                'JetbH_p4',
+                f'ConstructP4({feature_names[10]}, JetbH_eta, JetbH_phi, Take(Jet_mass_cut, bH_idx))' 
+            )            
+    )
+
+    # # build features 8 other features
+    df2 = (
+
+        df2.Define(
+            f'{feature_names[0]}',
+            'sqrt(pow(Lepton_eta.at(0)-JetbL_eta,2.)+pow(Lepton_phi.at(0)-JetbL_phi,2.))'
+        ).Define(
+            f'{feature_names[1]}',
+             'sqrt(pow(JetW1_eta-JetW2_eta,2.)+pow(JetW1_phi-JetW2_phi,2.))'
+        ).Define(
+            f'{feature_names[2]}',
+             'sqrt(pow(JetW1_eta-JetbH_eta,2.)+pow(JetW1_phi-JetbH_phi,2.))'
+        ).Define(
+            f'{feature_names[3]}',
+             'sqrt(pow(JetW2_eta-JetbH_eta,2.)+pow(JetW2_phi-JetbH_phi,2.))'
+        ).Define(
+            f'{feature_names[4]}', 'return Map(JetbL_p4+Lepton_p4.at(0), [] (const ROOT::Math::PxPyPzMVector &p) {return p.M();})'
+        ).Define(
+            f'{feature_names[5]}', 'return Map(JetW1_p4+JetW2_p4, [] (const ROOT::Math::PxPyPzMVector &p) {return p.M();})'
+        ).Define(
+            f'{feature_names[6]}', 'return Map(JetW1_p4+JetW2_p4+JetbH_p4, [] (const ROOT::Math::PxPyPzMVector &p) {return p.M();})'
+        ).Define(
+            f'{feature_names[7]}', 'return Map(JetW1_p4+JetW2_p4+JetbH_p4, [] (const ROOT::Math::PxPyPzMVector &p) {return p.Pt();})'
+        )
+    ) 
+    # put all features into one collection
+    feature_names = feature_names.__str__().replace("[","{").replace("]","}").replace("'","")
+    df2 = df2.Define("features",  
+        f"ROOT::VecOps::RVec<ROOT::RVecF>({feature_names})"
+        )
+
+    return df2
+
+
+    
+
+def predict (df: ROOT.RDataFrame)->ROOT.RDataFrame:
+
+    # get indexes of best permutations' candidates
+    return df.Define("perm_idx", 
+        """
+        bool is_even = (event % 2 == 0);
+        const auto& forest = (is_even) ? feven : fodd;
+        auto result  = inference(features, forest, true);
+        return ArgMax(result);
+        """
+    )
+
+def infer (df: ROOT.RDataFrame)->ROOT.RDataFrame:
```

Comment:
function name is too generic

Reply:
Why? To function "infer" is delegated everything about performing the inference. So, there is not more precise name, I think

Reply:
The problem is how this looks for someone reading `analysis.py`: `df4j2b = infer(df4j2b)`.

We want a function name that gives the reader as much information as possible about how the function changes the RDF object. For example `df = define_output_ml_features(df)`.

Reply:
add comment and documentation?